### PR TITLE
chore(integration-karma): avoid conditional tests

### PR DESCRIPTION
**Pull request by cardoso**
## Details

Currently the karma tests use if/else statements to decide which tests to run under various conditions.

The problem with this approach is that the total number of tests is not consistent and the number of tests executed and skipped is not accurate:

[]
```
Executed 3369 of 3695 (skipped 326) SUCCESS (19.83 secs / 17.888 secs)
```
[DISABLE_SYNTHETIC=1]
```
Executed 3205 of 3531 (skipped 326) SUCCESS (5.412 secs / 4.206 secs)
```
[API_VERSION=60]
```
Executed 3335 of 3661 (skipped 326) SUCCESS (20.422 secs / 18.118 secs)
```
[DISABLE_SYNTHETIC=1, API_VERSION=60]
```
Executed 3156 of 3482 (skipped 326) SUCCESS (5.118 secs / 3.999 secs)
```

This makes it hard to reason about the effects of the various flags and their combinations in dev and CI.

Knowing how many tests exist in total and how many are skipped under certain conditions will be especially important when migrating from karma (#3589)  to a different tool.

This PR proposes using the `vitest/no-conditional-tests` rule in karma tests.

It also adds an implementation of `{describe/it}.{runIf/skipIf}` inspired by vitest to be used instead of if/else.

After these changes:

[]
```
Executed 3369 of 4683 (skipped 1314) SUCCESS (19.35 secs / 17.499 secs)
```
[DISABLE_SYNTHETIC=1]
```
Executed 3205 of 4683 (skipped 1478) SUCCESS (5.689 secs / 4.24 secs)
```
[API_VERSION=60]
```
Executed 3335 of 4683 (skipped 1348) SUCCESS (20.269 secs / 17.862 secs)
```
[DISABLE_SYNTHETIC=1, API_VERSION=60]
```
Executed 3156 of 4683 (skipped 1527) SUCCESS (5.669 secs / 4.183 secs)
```

I wrote a script to run every non-hydration command used in the karma.yml workflow file and all reported 4683 total tests. The hydration tests don't have this issue as it always runs a single test per file (136 total).

I expected a downside of increased test duration, but thankfully I can't see anything noticeable.

I tried to keep the diff small with some indentation / formatting effort, but it still came up a bit large.

## Does this pull request introduce a breaking change?

-   😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

-   🤞 No, it does not introduce an observable change.

### CODE REVIEW

File: packages/@lwc/integration-karma/helpers/test-setup.js [hunk 1]
```diff
@@ -116,5 +116,21 @@ afterAll(function () {
...
+};
+
+it.runIf = function (condition) {
+    return condition ? it : xit;
+};
+
+it.skipIf = function (condition) {
+    return condition ? xit : it;
+};
+
```
Comment (nolanlawson) [inline packages/@lwc/integration-karma/helpers/test-setup.js:134]: This is a really neat solution!

**Review by nolanlawson — COMMENTED** [review]
Overall I'm in favor of this. It would be great to have a consistent # of tests between runs, and the Vitest-on-Jasmine shim code here is really minimal.

I'll give this a more thorough review later, but I'm supportive of this!

**Comment by cardoso** [comment]
Thanks @nolanlawson ! I see one test failed. I'll fix it as soon as I get back home.

That can happen if the code runs inside the describe block (outside any hook or test case) during collection of tests.

I fixed what I could catch, but LEGACY_BROWSERS=1 I hoped would just work 😆 . How do I reproduce this flag locally?

**Comment by nolanlawson** [comment]
Unfortunately you cannot reproduce it... unless you have some super-old browsers installed somehow. 🙂 

Don't sweat it; I can fix these tests myself as I review this PR... will get to it this week hopefully.

File: packages/@lwc/integration-karma/helpers/test-setup.js [hunk 2]
```diff
@@ -116,5 +116,21 @@ afterAll(function () {
     throwIfConsoleCalled();
 });
 
+describe.runIf = function (condition) {
```
Comment (nolanlawson) [inline packages/@lwc/integration-karma/helpers/test-setup.js:119]:
```suggestion
// The following (`runIf`, `skipIf`, etc.) are based on Vite's APIs: https://vitest.dev/api/
// This allows us to use the vitest/no-conditional-tests ESLint rule and get the same total # of tests for
// every variant of a test run (e.g. `DISABLE_SYNTHETIC=1`, `NODE_ENV_FOR_TEST=production`, etc.)
describe.runIf = function (condition) {
```

**Comment by nolanlawson** [comment]
Code coverage is the same before and after

Screenshot 2024-10-17 at 2 39 53 PM

**Review by nolanlawson — APPROVED** [review]
Looks good! Thank you!!

**Pull request merged**